tony-core/src/main/java/com/linkedin/tony/TonyApplicationMaster.java

```diff
@@ -427,14 +428,19 @@ private boolean prepare() {
   }
 
   /**
-   * Create job directory under history folder.
+   * Create job directory under specific date folder that job is carried out.
    * @param fs FileSystem object.
    * @param histFolder History folder location string.
    * @param appId Application ID string.
    * @throws IOException when failed to create job directory on HDFS
    */
-  private void setupJobDir(FileSystem fs, String histFolder, String appId) throws IOException{
-    jobDir = new Path(histFolder + "/" + appId + "/");
+  private void setupJobDir(FileSystem fs, String histFolder, String appId) throws IOException {
+    Calendar cal = Calendar.getInstance();
+    int yr = cal.get(Calendar.YEAR);
+    int mo = cal.get(Calendar.MONTH) + 1;
+    int d = cal.get(Calendar.DATE);
+    String jobFilePath = String.format("%s/%d/%d/%d/%s", histFolder, yr, mo, d, appId);
```

Comment:
I realized there's an issue with having the user's application create the `yyyy/mm/dd` directories. If a user's app creates these directories, other users won't be able to write to them, as the directory will be owned by the user who created the directories initially and the directories won't be world-writable (we don't want them to be world-writable).

So I think we need to have separate `intermediate/` and `finished` directories just like MR has. The `intermediate/` directory can have `<username>` subdirectories.

It should be the TonY History Server that reads all the `intermediate/` folders and moves them to the `finished/` directory, creating the `yyyy/mm/dd` directories under `finished/` if needed.

Reply:
Also, I realized that the old code didn't really set the permission for parent directories, but just the last component. I fixed that in this commit, with 770 permission. https://github.com/linkedin/TonY/pull/104/commits/f922881c006da30cbef51f001261d054ebc0b1c4.

Reply:
That's a possible solution, but one downside is by making the folder `0770`, any user can now modify/delete/overwrite any other user's config files. We also don't want the user who creates the yyyy/mm/dd directory to be able to just recursively delete the entire folder either.

Reply:
Let's say we were to split the jobs to intermediate/finished folders, and THS takes care of moving jobs from `intermediate` to their respective folders in `finished`. The job folder in intermediate has to be set to sth like `0750` (since we want to avoid the problem of any specific user has the capability to recursively delete others' files). Then we have another problem where a user the only person that can remove their job directories, and THS won't be able to clean up the job directories in `intermediate` due to permission issue. Doing `ls` on `tony-history/intermediate` will result in memory limit exceeded, like what happened to Spark history folder.

Reply:
The folders in `intermediate` can be set with `770` permissions and group `tony`, so that only the user who ran the job and users in the tony group (should only be tony + some admins) can move/delete the directories/files.

Reply:
👍